book/part1/project-lifecycles.md

```diff
@@ -1 +1,191 @@
+
+First is necessary define a problem to solve: *What are we going to do?*
+
+Two techniques are useful in the Ideation step:
+
+* Brainstorming
+* Research
+
+```{admonition} Brainstorming
+Brainstorming helps to explore your project concepts, generate new ideas, and 
+reject bad ones.
+```
+
+```{admonition} Research
+To create a relevant work it is necessary to know the environment in which it is
+inserted. Research helps to validate ideas from brainstorming sessions
+```
+
+## Requirements
+
+Define a project plan: *How are we going to do it?*
+
+In this step it is necessary to gather software requirements from **everyone**
+involved in the project.
+
+```{warning}
+Each person involved in the project may have a different need, so it is
+necessary to check with each one what they need. 
+
+For example, in an academic environment, the PI usually has a more global view
+of the project, while other participants have a more specific view.
+```
+
+Some topics to help define requirements include:
+
+* final goal
+* project scope (how to reach the final goal)
+* what is feasible (and how)
+* what is priority
+* what resources are available
+* deadlines
+* potential risks
+
+### Software Requirement Specification
+
+All information regarding the requirements must be organized in a file, the
+*Software Requirement Specification*.
+
+**Everyone** involved in the project should review this document.
+If any requirement has a problem (is incomplete, unclear, ambiguous, etc.), it
+must be identified and corrected.
+
+```{tip}
+Depending on the characteristics of the project, it may be useful to generate
+some additional documents to help understand the requirements.
+
+For example, if the project is related to data collection, a Data Flow Diagram
+can be help to track all variables, their connections and clarify requirements.
+```
+
+## Design
+
+Translate the Software Requirement Specification into Design: *What is the
+software architecture?*
+
+To design software, the *object-oriented* approach is the predominant
+programming paradigm.
+
+```{admonition} Object-oriented components
+* **Classes**: A user-defined type
+* **Object instances**:  A particular object instantiated from a class.
+* **Methods**: A function which is “built in” to a class
+* **Constructor**: A special method called when instantiating a new object
+```
+
+The design usually includes descriptions of the architecture overview (high
+level design) and the design of its components (low-level design).
+
+```{tip}
+When design an object-oriented software you must follow some principles, like:
+
+* Abstraction
+* Encapsulation
+* Decomposition
+* Generalisation
+```
+
+The design can be described using a class diagram in the Unified Modelling
+Language (UML). See below an UML class diagram for a purchasing system software:
+
+![Class Diagram Example: Order System](https://cdn-images.visual-paradigm.com/guide/uml/uml-class-diagram-tutorial/17-class-diagram-example-order-system.png)
```

Comment:
Might be worth adding a line of attributing to this diagram
`This figure was adapted from/originally used by ... at https://webwebweb.com`